Enhancements to Account Management and Introducing Image Handling

Dragonfly911117 opened the pull request:
# Title
Enhancements to Account Management and Introducing Image Handling

# Description
This pull request introduces significant enhancements to account management, including new models and improved CRUD operations, along with the adoption of new functionality for image uploads and retrievals. Here's a breakdown of the changes introduced:

1. **Account Management**
   Refined the CRUD operations for accounts (`router/account.py`). The changes in `model/account.py` include new models for handling account data with additional fields.

2. **Image Handling**
   A new feature has been introduced to handle image uploads and retrievals. New routes have been introduced in `router/image.py` and the corresponding models in `model/image.py`. A helper file, `utils/image_io.py` has been added to handle the related operations.

3. **Database Processes**
   `utils/db_process.py` has been modified and includes a new function to check if a record exists in a given database table.

4. **Shop Module Enhancement**
   Minor improvements and cleanups have been made to the Shop router (`router/shop.py`).

5. **Dependencies**
   The `requirements.txt` file has been updated to list specific versions of the dependencies to ensure consistent behavior across different environments.

6. **Server Setup**
   Changes in `server.py` reflect the updates in routing rules, factoring in the new routes introduced in this PR.

Please note: It is essential to thoroughly test these changes in a staging or development environment before rolling out to production. These changes are interdependent and should be deployed together.

Any feedback or questions are appreciated as we aim to continually improve our codebase.

Dragonfly911117 committed:
Delete example account router and cleanup things

The router/account.py had unnecessary imports and an endpoint that seemed to serve only as an example, with a hardcoded success message, so it has been deleted. In the core/settings.py and model/account.py, trailing spaces were removed to improve the code readability. In the main.py, the spaces around the assignment operators have been removed as our style guide prefers no spaces around operators in function arguments.

Dragonfly911117 committed:
Add image_io module for image uploading

This commit introduces a new module, image_io.py, that is used for image uploading. The module has functionality to list files, get directory path, generate a unique file name for uploads, and save the uploaded file in the correct location. The implementation currently supports differentiating between avatar and product uploads. This module will provide the necessary underlying support for any image uploading requirements in the application.

Dragonfly911117 committed:
Add product image upload feature and refactor

Implemented a product image upload feature that restricts file types to jpeg or png and limits the file size to 10MB. Renamed account to product for better clarity of use case. Included python-multipart in the requirements for file uploading capabilities. Refactored the save_file method to be asynchronous and included the product id in the directory structure for organisation and easy retrieval.

Dragonfly911117 committed:
Add account router and refactor image upload

Account router was added to the server.py to provide endpoint for account-related operations. The image file validation and size restrictions were moved to save_file function in image_io.py to avoid duplication and enhance consistency across different routers. Now, not only product but also account router reuses this function for image uploads, ensuring that the same validation and restrictions are applied.

Dragonfly911117 committed:
Update error messages to include actual value in image_io.py

Modified http_error exceptions to include the actual values in their respective error message for better debugging and user understanding in image_io.py. Now, when validation fails for file content type or file size, the error message will specify the received value.

Dragonfly911117 committed:
Optimize routing and file handling

Restructured the routes for product and account into separate modules (product.py and account.py respectively). Image upload functionality moved to a separate module to segregate responsibilities for cleaner, more organized code. This also results in a more intuitive directory structure. The server port has also been changed from 8090 to 8000 for standardization. Imported utility functions have also been restructured to be contained in a tool specific file (image_io.py). This enhances code readability and boosts efficient execution.

Dragonfly911117 committed:
Merge branch 'master' into image

# Conflicts:
#	model/account.py
#	requirements.txt
#	router/account.py
#	server.py

Dragonfly911117 committed:
Refactor account-related code and restructure import sequence

Reorganized imports in 'model/account.py' for cleaner code structure. Migrated account-related functionalities into separate files (create.py, get.py, update.py) under 'router/Account/' directory facilitating better code organization and readability. 'router/Account/account.py' now includes a function to gather all routers from the different files within 'Account' directory. Redundant imports were removed from 'server.py' and 'router/entries.py'. Finally, the file 'router/account.py’ was renamed to 'router/Account/create.py', 'router/shop.py' was renamed to 'router/Shop/shop.py' to maintain a unified coding style.

Dragonfly911117 committed:
Refactor router structure for cleaner codebase

In this commit, the structure for FastAPI routers has been consolidated and simplified to better accommodate future development. Specifically, the previously separate account related router files (create.py, get.py, update.py, image.py and account.py) have been merged into a single file named account.py, with router becoming a top-level object. This refactoring makes the codebase much cleaner and easier to navigate. The same approach was applied for the product related routers. Changes were also made in entries.py to adapt to the new file structure. As for functionality, everything remains the same. This change was necessary for maintaining an organized codebase and allowing easier addition of new APIs in the future.

Dragonfly911117 committed:
Refactor router import and error handling in APIs

The major change in this commit is to refactor the import statement in server.py to simplify the codebase. Instead of importing from 'router.entries' it now directly imports from 'router'. Furthermore, some changes were done in the exception handling of image_io.py, removing the explicit error messages for now. Lastly, type hints were added to upload_image and get_image APIs in account.py where 'id' is now explicitly treated as a string. .

Dragonfly911117 committed:
Improve error handling in save_file and related APIs

Refactored error handling mechanism in save_file function in image_io.py to raise exceptions instead of returning boolean values. This resulted in changes in upload_image APIs in account.py and product.py where these exceptions are now caught and appropriate responses are returned. Also, save_file now returns filename rather than a boolean. Consequently, in ImageIOSuccessModel, "filename" field is replaced by "file_name". Furthermore, "id" in upload_image and get_image APIs in account.py is now type hinted as a string.

Dragonfly911117 committed:
Refactor image file handling and upgrade dependencies

This commit refactors the way image files are handled while being saved. The HTTP exceptions were replaced by JSON responses directly in the `save_file` function, improving code reusability and reducing duplication in `account.py` and `product.py`. By returning JSON responses directly, it improves transparency and provides clearer feedback to the end users about the status and possible errors. Dependencies in `requirements.txt` were also upgraded to their latest stable versions to align with latest features and security updates. Type hint for the variable 'id' was also added in the image upload APIs for more explicit error handling.

Dragonfly911117 committed:
Add check for account/product existence before uploading image

Before saving an image for an account or product, the existence of that account or product is now determined. This change was made to handle cases where a non-existing ID might be used as an argument in the upload_image functions. If the account or product does not exist, the functions will return a JSON response with a status of 400 (Bad Request) and a respective error message, improving the API's error handling. This will help prevent erroneous data being associated with non-existent IDs and provide a clearer feedback to the client if such a mistake is made.

Dragonfly911117 committed:
Rearrange route paths for image uploading and fetching

The route paths for uploading and fetching images in the account and product modules have been rearranged to improve readability and consistency in the API interface. Previously, the ID was positioned before the action in the path (e.g., "/{id}/upload_image"), which could lead to confusion. Now, the action comes first, followed by the ID (e.g., "/upload_image/{id}"), making the paths more intuitive and reflective of the actions being taken. This rearrangement conforms more to RESTful principles. No functionality changes were made.

Dragonfly911117 committed:
Improve result fetching and file handling methods

Refactored the `get_all_result` method to `get_all_results` in `db_process.py` for more accurate semantic representation. Introduced a `get_file` method in `image_io.py` to enable file fetching, enhancing the system's file management functionality. Reorganized import statements in several modules for improved readability. All referring calls to `get_all_results` have been updated in accordance.

Dragonfly911117 committed:
Refactor image handling and database result fetching methods

Reworked methods to enhance the system's efficiency and readability. `get_all_result` method was renamed to `get_all_results` in `db_process.py` signifying a more accurate depiction of the method's function. Added a `get_file` method in `image_io.py` to facilitate file fetching, boosting the system's file management feature. Import statements in several modules were reshuffled for optimizing readability. All calls referring to `get_all_results` have been updated to align with these changes.

Dragonfly911117 committed:
Fix sussy formatting issues

Dragonfly911117 committed:
"Add feature to download all images as zip

A new capability for downloading all images of a specific owner as a zipfile has been added. For this, the 'get_all_files' function was introduced in 'image_io.py', which archives all the images in a zipfile and returns it as a streaming response. The endpoint for this feature was added in 'image.py' under the route '/get_all_images/{owner_id}/all', which uses the aforementioned function. This feature was required to let users bulk download all of their images at once."

Dragonfly911117 committed:
Merge branch 'master' into image

Dragonfly911117 committed:
Refactor code for better readability

The changes consist of reformatting the function decorators to match a standard Python multi-line function decorator formatting. Function calls and nested data structures were also broken down into multiple lines for more readability. Additionally, nondescript variable names like 'id' were changed to more meaningful names like 'account_id' or 'shop_id'. This should help in understanding the code and function arguments for better code navigation and cooperation between team members.

Dragonfly911117 committed:
Refactor tags and decorator formatting in multiple routers

Updated tags and refactored function decorators in the image, shop, and account routers. The changes enhance readability by making tags more descriptive and aligning with standard conventions for multi-line function decorators. This modification will aid in better code navigation and cooperation.

Vghxv commented on router/account.py:
```diff
@@ -13,7 +13,7 @@
 
 @router.post(
-    "/create", tags=["create"], responses={
+    "/", tags=["create", "HTTP_POST"], responses={
```
tag "create" and "HTTP_POST" is the same concept, remove one of them so swagger looks cleaner, the same thing applies to get/HTTP_GET and update/HTTP_PUT

PEKKA1117 replied:
fixed with 973f9d6

Vghxv commented on router/account.py:
```diff
@@ -1,125 +1,140 @@
-    account_form: CreateAccountForm = Depends(CreateAccountForm.as_form)
+)
+async def update_account(
+        account_form: UpdateAccountForm = Depends(UpdateAccountForm.as_form)
```
sussy spaces, check all routers input

Dragonfly911117 committed:
Remove redundant HTTP method tags from routers

Removed the "HTTP_GET", "HTTP_POST" and "HTTP_PUT" tags from the router decorators in `shop.py`, `account.py` and `image.py`. These tags were redundant and unnecessary as the HTTP method is already specified in the decorator itself (@router.get, @router.post, @router.put). This simplifies our tag system and makes the code cleaner and easier to read.

Dragonfly911117 committed:
Refactor image upload and retrieval process

Image uploads and retrieval have been enhanced to handle different types of images (avatar and banner). The ImageTypeModel enum class has been added to "model/image.py". The function 'if_exists_in_db' in "utils/db_process.py" became asynchronous to facilitate the new changes. The 'upload_image' function in "router/image.py" now accepts an 'img_type' parameter to tailor its behaviour based on the provided image type. High-level changes were made in "utils/image_io.py" to reflect these refactoring. These changes should allow for more functionality and flexibility when dealing with uploads and retrievals of different image types.

Dragonfly911117 committed:
Update ImageUploadForm and ImageTypeModel in model/image.py

The ImageUploadForm and ImageTypeModel in the image.py module have been updated to simplify and improve the image upload and retrieval process. ImageUploadForm, which previously handled image uploads, and ImageGetForm, which dealt with image retrievals, have been removed. The ImageTypeModel enum class has been introduced to differentiate between two types of images - avatars and banners. The error message for ImageIOFail model has been updated to make it clearer. These changes will allow a more flexible and functional interaction with different image types.

Dragonfly911117 committed:
Change error message in ImageIOFailModel class

Updated the default error message in the ImageIOFailModel class from 'Failed to upload image!' to 'Failed'. This is to make the error message more general since the ImageIOFailModel class is used not just for image upload failures

Dragonfly911117 committed:
Simplify image upload endpoint URL

The URL for the image upload endpoint in the router was simplified from "/upload/{owner_id}/{img_type}" to "/upload". This should simplify interaction with the endpoint, as it no longer requires unnecessary dynamic segments. It's a part of API refactoring for better usability.
